File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>
#include<iostream>
using namespace std;

int vis[101][101];
int dx[4] = {1,-1,0,0};
int dy[4] = {0,0,1,-1};

int solution(vector<vector<int>> maps){
    int answer = -1;
    queue<pair<int,int>> q;
    int x = maps.size()-1;
    int y = maps[0].size()-1;
    
    for(int i=0; i<x; i++){
        for(int j=0; j<y; j++){
            vis[i][j] = 0;
        }
    }
    
    vis[0][0] = 1;
    q.push({0,0});
    
    while(!q.empty()){
        pair<int,int> cur = q.front(); q.pop();
        if(cur.first == maps.size()-1 && cur.second == maps[0].size()-1){
            answer = vis[cur.first][cur.second];
            return answer;
        }
    
        for(int dir=0; dir<4; dir++){
            int nx = cur.first + dx[dir];
            int ny = cur.second + dy[dir];
            
            if(nx<0 || nx>x || ny<0 || ny>y) continue;
            if(vis[nx][ny]==0 && maps[nx][ny]){
                vis[nx][ny] = vis[cur.first][cur.second]+1;
                q.push({nx,ny});
            }
        }
    }
    return answer;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    class Cur {
        int x;
        int y;

        public Cur(int x,int y){
            this.x = x;
            this.y = y;
        }
    }   
    public int solution(int[][] maps) {
        int n = maps.length-1;
        int m = maps[0].length-1;
        int answer = 0;
        int[][] vis = new int[n+1][m+1];
        
        bfs(maps,vis,n,m);
        answer = vis[n][m];
        
        if(answer == 0){
            return -1;
        }
        
        return answer;
    }
    
    private void bfs(int[][] maps, int[][] vis, int n, int m){
        int[] dx = {0,1,0,-1};
        int[] dy = {1,0,-1,0};
        
        Queue<Cur> q = new LinkedList<>();
        q.add(new Cur(0,0));
        vis[0][0] = 1; 
        
        while(!q.isEmpty()){
            Cur cur = q.poll();

            for(int i=0; i<4; i++){
                int nx = dx[i] + cur.x;
                int ny = dy[i] + cur.y;

                if(nx>=0 && nx<=n && ny>=0 && ny<=m){
                    if(vis[nx][ny] == 0 && maps[nx][ny] == 1){
                        q.add(new Cur(nx,ny));
                        vis[nx][ny] = vis[cur.x][cur.y] + 1;
                    }
                }
            }
        }
        
    }
}